util/webhook/validator.go

```diff
@@ -0,0 +1,141 @@
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// Validator defines functions for validating an operation.
+type Validator interface {
+	runtime.Object
+	ValidateCreate(client client.Client) error
+	ValidateUpdate(old runtime.Object, client client.Client) error
+	ValidateDelete(client client.Client) error
+}
+
+// NewValidatingWebhook creates a new Webhook for validating the provided type.
+func NewValidatingWebhook(validator Validator, client client.Client) *admission.Webhook {
+	return &admission.Webhook{
+		Handler: &validatingHandler{
+			validator: validator,
+			Client:    client,
+		},
+	}
+}
+
+type validatingHandler struct {
+	validator Validator
+	Client    client.Client
+	decoder   *admission.Decoder
+}
+
+var _ admission.DecoderInjector = &validatingHandler{}
+
+// InjectDecoder injects the decoder into a validatingHandler.
+func (h *validatingHandler) InjectDecoder(d *admission.Decoder) error {
+	h.decoder = d
+	return nil
+}
+
+// Handle handles admission requests.
+func (h *validatingHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	if h.validator == nil {
+		panic("validator should never be nil")
+	}
+
+	// Get the object in the request
+	obj := h.validator.DeepCopyObject().(Validator)
+	if req.Operation == v1.Create {
+		err := h.decoder.Decode(req, obj)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		err = obj.ValidateCreate(h.Client)
+		if err != nil {
+			var apiStatus apierrors.APIStatus
+			if goerrors.As(err, &apiStatus) {
+				return validationResponseFromStatus(false, apiStatus.Status())
+			}
+			return admission.Denied(err.Error())
+		}
+	}
+
+	if req.Operation == v1.Update {
+		oldObj := obj.DeepCopyObject()
+
+		err := h.decoder.DecodeRaw(req.Object, obj)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+		err = h.decoder.DecodeRaw(req.OldObject, oldObj)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		err = obj.ValidateUpdate(oldObj, h.Client)
+		if err != nil {
+			var apiStatus apierrors.APIStatus
+			if goerrors.As(err, &apiStatus) {
+				return validationResponseFromStatus(false, apiStatus.Status())
+			}
+			return admission.Denied(err.Error())
+		}
+	}
+
+	if req.Operation == v1.Delete {
+		// In reference to PR: https://github.com/kubernetes/kubernetes/pull/76346
+		// OldObject contains the object being deleted
+		err := h.decoder.DecodeRaw(req.OldObject, obj)
+		if err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+
+		err = obj.ValidateDelete(h.Client)
+		if err != nil {
+			var apiStatus apierrors.APIStatus
+			if goerrors.As(err, &apiStatus) {
+				return validationResponseFromStatus(false, apiStatus.Status())
+			}
+			return admission.Denied(err.Error())
+		}
+	}
+
+	return admission.Allowed("")
+}
+
+func validationResponseFromStatus(allowed bool, status metav1.Status) admission.Response {
+	resp := admission.Response{
```

Comment:
nit, can just return the struct directly

Reply:
ack